WebDriverAgentLib/Utilities/FBImageProcessor.m

```diff
@@ -42,86 +43,118 @@ - (id)init
-// and should be used in `submitImage` API, while the `scaledImageWithImage`
+// and should be used in `submitImage` API, while the `scaledImageWithData`
 // one is more generic
-- (nullable NSData *)scaledJpegImageWithImage:(NSData *)image
-                                scalingFactor:(CGFloat)scalingFactor
-                           compressionQuality:(CGFloat)compressionQuality
-                                        error:(NSError **)error
++ (nullable NSData*)processedJpegImageWithData:(NSData *)imageData
+                                 scalingFactor:(CGFloat)scalingFactor
+                                         error:(NSError **)error
 {
-  CGImageSourceRef imageData = CGImageSourceCreateWithData((CFDataRef)image, nil);
-  CGSize size = [self.class imageSizeWithImage:imageData];
-  CGFloat scaledMaxPixelSize = MAX(size.width, size.height) * scalingFactor;
-  CFDictionaryRef params = (__bridge CFDictionaryRef)@{
-    (const NSString *)kCGImageSourceCreateThumbnailWithTransform: @(YES),
-    (const NSString *)kCGImageSourceCreateThumbnailFromImageIfAbsent: @(YES),
-    (const NSString *)kCGImageSourceThumbnailMaxPixelSize: @(scaledMaxPixelSize)
+  CGImageSourceRef imageDataRef = CGImageSourceCreateWithData((CFDataRef)imageData, nil);
+  
+  NSDictionary *options = @{
+    (const NSString *)kCGImageSourceShouldCache: @(NO)
   };
-  CGImageRef scaled = CGImageSourceCreateThumbnailAtIndex(imageData, 0, params);
-  CFRelease(imageData);
-  if (nil == scaled) {
-    [[[FBErrorBuilder builder]
-      withDescriptionFormat:@"Failed to scale the image"]
-     buildError:error];
-    return nil;
+  CFDictionaryRef properties = CGImageSourceCopyPropertiesAtIndex(imageDataRef, 0, (CFDictionaryRef)options);
+  NSNumber *width = [(__bridge NSDictionary *)properties objectForKey:(const NSString *)kCGImagePropertyPixelWidth];
+  NSNumber *height = [(__bridge NSDictionary *)properties objectForKey:(const NSString *)kCGImagePropertyPixelHeight];
+  CGImagePropertyOrientation orientation = (CGImagePropertyOrientation) [[(__bridge NSDictionary *)properties objectForKey:(const NSString *)kCGImagePropertyOrientation]
+                                                                         integerValue];
+  CGSize size = CGSizeMake([width doubleValue], [height doubleValue]);
+  CFRelease(properties);
+
+  BOOL usesScaling = scalingFactor > 0.0 && scalingFactor < FBMaxScalingFactor;
+  CGImageRef resultImage = NULL;
+  if (orientation != kCGImagePropertyOrientationUp || usesScaling) {
+    CGFloat scaledMaxPixelSize = MAX(size.width, size.height) * scalingFactor;
+    CFDictionaryRef params = (__bridge CFDictionaryRef)@{
+      (const NSString *)kCGImageSourceCreateThumbnailWithTransform: @(YES),
+      (const NSString *)kCGImageSourceCreateThumbnailFromImageIfAbsent: @(YES),
+      (const NSString *)kCGImageSourceThumbnailMaxPixelSize: @(scaledMaxPixelSize)
+    };
+    resultImage = CGImageSourceCreateThumbnailAtIndex(imageDataRef, 0, params);
```

Comment:
This is a very CPU expensive operation to call if orientation change is required. During debugging on my iPhone 13 Pro Max I saw that the CPU usage jumps to a constant 95% while the orientation is not portrait. By repeating the test @KazuCocoa ran with Chrome on my device I noticed that when in portrait WDA is stopped by the OS due to the increased CPU usage; an excessive CPU usage report is generated stating "48 seconds CPU time over 57 seconds (84% cpu average), exceeding limit of 80% cpu over 60 seconds".

Reply:
>  when in portrait 

you meant in landscape?

> increased CPU usage

I cannot find any other APIs to fix both scale and rotation. This one seems to be the most optimised one. Do you know any alternative APIs that could use GPU acceleration, so it does not overload the CPU?

Reply:
Yes, I meant to say not in portrait, sorry.
Regarding the API you found - worth a try. I'll try to see if I can find an alternative which could utilize hardware acceleration.

Reply:
According to [this](https://nshipster.com/image-resizing/) article the ImageIO approach seems to be one of the fastest

Reply:
@Dan-Maor I have changed the code to use UIGraphicsImageRenderer. Please check if it makes any difference in performance.

Reply:
@mykola-mokhnach unfortunately this appears to be a bit worse, when rotating to landscape aside from the increased CPU payload the memory allocation is increasing very quickly and is not releasing until the device is rotated back to portrait.
Also - when the device orientation is landscape left the image is not fully displayed when testing the video stream in Chrome:
![Clipboard01](https://github.com/appium/WebDriverAgent/assets/24937888/5dbf0e75-d728-480b-ba5c-cf5179cfd7c4)

Reply:
Yea, it seems like https://github.com/appium/WebDriverAgent/pull/812/files#diff-1dc7ca4b743bd01948c46ff0f17376c7689fb62f16aa94a10d0338a6b8c0785fR123-R138 keep increasing the memory usage 15 MB - 20MB every time, then the WDA process was stopped after reaching 1.5GB and a bit.

Reply:
I haven't verified well yet, but wrapping with `@autoreleasepool` may help to prevent the memory usage itself for https://github.com/appium/WebDriverAgent/pull/812/files#diff-1dc7ca4b743bd01948c46ff0f17376c7689fb62f16aa94a10d0338a6b8c0785fR121-R139

Reply:
This method also does not seem very reliable to me. It seems like the image gets rendered differently on simulator and on real devices. The ImageIO-based one worked the best so far

Reply:
I've changed it again to use the https://developer.apple.com/documentation/uikit/uiimage/3750845-preparethumbnailofsize?language=objc API. It is much simpler to use, not sure about the performance though